### PR TITLE
[aot] Instead of decoding debug info in decode_exception_debug_info (), save the location of it, and decode it when its needed by mono_debug_find_method ().

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mono/metadata/mono-debug.c [hunk 1]
```diff
@@ -710,20 +718,31 @@ find_method (MonoMethod *method, MonoDomain *domain, MonoDebugMethodJitInfo *jit
...
 
-	if (!address)
-		return NULL;
+	if (!address) {
+		/* Load on demand from AOT images */
+		if (find_jit_debug_info_cb) {
+			MonoDebugMethodJitInfo *tmp = find_jit_debug_info_cb (domain, method);
+			if (tmp) {
+				/* The caller expects the result to be copied to @jit */
+				memcpy (jit, tmp, sizeof (MonoDebugMethodJitInfo));
```
Comment (jaykrell) [inline mono/metadata/mono-debug.c]: You can just say `*jit = *tmp` in ANSI C.

File: mono/mini/debug-mini.c [hunk 2]
```diff
@@ -616,13 +621,18 @@ mono_debug_add_aot_method (MonoDomain *domain, MonoMethod *method, guint8 *code_
...
+	/* Cache the data here, and decode it when its needed in mini_debug_find_jit_debug_info () */
+	AotDebugInfo *info = mono_domain_alloc0 (domain, sizeof (AotDebugInfo));
+	info->code_start = code_start;
+	info->debug_info = debug_info;
+	info->debug_info_len = debug_info_len;
+
+	domain_info = domain_jit_info (domain);
+	mono_domain_lock (domain);
+	if (!domain_info->aot_debug_info_hash)
+		domain_info->aot_debug_info_hash = g_hash_table_new_full (mono_aligned_addr_hash, NULL, NULL, g_free);
```
Comment (lateralusX) [inline mono/mini/debug-mini.c]: Should this free memory returned by mono_domain_alloc0?

Comment (vargaz) [reply mono/mini/debug-mini.c]: Fixed.

**Comment by jaykrell** [comment]
C++
`debug-mini.c:785:30: error: invalid conversion from 'gpointer {aka void*}' to 'AotDebugInfo*' [-fpermissive]`

File: mono/mini/debug-mini.c [hunk 3]
```diff
@@ -616,13 +621,18 @@ mono_debug_add_aot_method (MonoDomain *domain, MonoMethod *method, guint8 *code_
...
 
-	jit = deserialize_debug_info (method, code_start, debug_info, debug_info_len);
-
-	mono_debug_add_method (method, jit, domain);
-
-	mono_debug_add_vg_method (method, jit);
-
-	mono_debug_free_method_jit_info (jit);
+	/* Cache the data here, and decode it when its needed in mini_debug_find_jit_debug_info () */
+	AotDebugInfo *info = mono_domain_alloc0 (domain, sizeof (AotDebugInfo));
```
Comment (jaykrell) [inline mono/mini/debug-mini.c]: I think this line -- needs a cast in C++.

File: mono/mini/debug-mini.c [hunk 4]
```diff
@@ -755,3 +765,28 @@ mono_debugger_method_has_breakpoint (MonoMethod *method)
...
+mini_debug_find_jit_debug_info (MonoDomain *domain, MonoMethod *method)
+{
+	MonoJitDomainInfo *domain_info;
+	MonoDebugMethodJitInfo *jit = NULL;
+	AotDebugInfo *info = NULL;
+
+	domain_info = domain_jit_info (domain);
+	mono_domain_lock (domain);
+	if (domain_info->aot_debug_info_hash)
+		info = g_hash_table_lookup (domain_info->aot_debug_info_hash, method);
```
Comment (jaykrell) [inline mono/mini/debug-mini.c]: Oh -- definitely this line, needs a cast=

File: mono/mini/debug-mini.c (hunk 3 above)
Comment (jaykrell) [reply mono/mini/debug-mini.c]: This line does not need a cast. It is the other one. We could make g_hash_table_lookup "auto cast" though.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
The locking is not right, needs a rework.

**Comment by vargaz** [comment]
Depends on:
https://github.com/mono/mono/pull/15367

**Comment by lewing** [comment]
@vargaz do you want to keep this open even though #15367 was closed?

**Pull request closed without merging**